src/population.py

```diff
@@ -1,58 +1,56 @@
         self.solutions = solutions
-        self.fitness = fitness
-
-    def initial_population(self, population_size, test_size):
-        population = list()
-        for i in range(population_size):
-            test = list()
-            for j in range(test_size):
-                test.append(random.randint(-1000, 1000))
-            population.append(test)
-        return population
-
-    def mutate_1(self, lower, upper, prob):
-        mutated = list()
-        for test in self.solutions:
-            test_list = list(map(int, test.split()))
-            for i in range(len(test_list)):
-                if random.random() < prob:
-                    test_list[i] = random.uniform(lower, upper)
-            test_mutated = "".join(test_list)
-            mutated.append(test_mutated)
-        return mutated
-
-    def mutate_2(self, lower, upper, prob, rate):
-        mutated = list()
-        for test in self.solutions:
-            test_list = list(map(int, test.split()))
-            for i in range(len(test_list)):
-                if random.random() < prob:
-                    test_list[i] = random.uniform(lower, upper)
-            test_mutated = "".join(test_list)
-            mutated.append(test_mutated)
-        prob -= rate
-        return mutated
-
-    def crossover(self, parent_1, parent_2, rate):
-        parent_1 = list(map(int, parent_1.split()))
-        parent_2 = list(map(int, parent_2.split()))
-        if random.random() < rate:
-            crossover_point = random.choice(range(len(parent_1)))
-            parent_1 = parent_1[:crossover_point] + parent_2[crossover_point:]
-            parent_2 = parent_2[:crossover_point] + parent_1[crossover_point:]
-        return parent_1, parent_2
-
-    def tournament_selection(self):
+
+    def initial_population(self, population_size: int, test_type: int):
+        for _ in range(population_size):
+            test = TestCase(test_type=test_type)
+            self.solutions.append(test.get_input())
+
+    def mutate(self, solution: TestCase, mode: int) -> TestCase:
+        offspring = copy.deepcopy(solution)
+        offspring[random.randint(0, len(solution) - 1)] = random.randint(0, 10000)
+
+        return offspring
+
+    def crossover(self, parent_1: TestCase, parent_2: TestCase) -> Tuple[TestCase]:
+        crossover_point = random.choice(range(len(parent_1)))
+        offspring_1 = parent_1[:crossover_point] + parent_2[crossover_point:]
+        offspring_2 = parent_2[:crossover_point] + parent_1[crossover_point:]
+
+        return offspring_1, offspring_2
+
+    def tournament_selection(self) -> Tuple[TestCase]:
         parents = []
         for _ in range(2):
-            x, y = random.choices(self.solutions)
-            if self.fitness[x] > self.fitness[y]:
-                parent = x
+            candidate_1, candidate_2 = random.choices(self.solutions)
+            if candidate_1.get_fitness() > candidate_2.get_fitness():
```

Comment:
I am not sure rn, but lower the fitness() -> better. bc constraint will be = 0 when it is satisfied. But let's see